### ISSUE TITLE
Support both /users/{user_id} and /users/_bulk

### ISSUE THREAD

**Issue by haizaar** [tiangolo/fastapi#244]
I have `/users` API that supports PUT and GET for `/users/{user_id}`. Now I want to introduce a bulk API to update several users at once. I thought about introducing `/users/_bulk` endpoint for PUT operations - we don't allow `_` in our IDs hence there is no URL collision.

However when trying to implement it in FastAPI they do collide and `/users/_bulk` gets routed to `/users/{user_id}`.

In our old system we use aiohttp which stores routes just a a list. Hence, to solve the above "collision" we just sort routes as
```
paths = sorted(paths, key=lambda x: x.count("{"))
```
which effectively pushes more vague, pattern routes down the list.

Do you think it will be a valid request for FastAPI to do the same?

**Comment by euri10** [comment]
Maybe not what you want, but have you tried changing the routes order?

**Comment by haizaar** [comment]
Yeap, I tried to put `@router.put("/users/_bulk", ...)` before `@router.put("/users/{user_id}")`, but it didn't help.
AFAIR, modules are still unordered dicts, even in Python3.6+.

**Comment by euri10** [comment]
Just mentioned that to you because if I 'm not mistaken routes are
evaluated orderly:
https://www.starlette.io/routing/

**Comment by haizaar** [comment]
Yeah, I saw that as well. And now I'm thinking that since python reads code top-down, it should've worked...

**Comment by euri10** [comment]
Another thing worth trying is Starlette urlconvertors

**Comment by euri10** [comment]
just tried and it seems the underscore makes the route not registered in the APIRouter

**Comment by haizaar** [comment]
It gets interesting :)
I had other issue that confused me to believe there is a problem. I re-read the docs you send and they confirm that my case should be supported with explicit ordering.

So now things work, **even regardless of the definition order!** I.e. exactly the same way I wanted originally.

I even tried `PUT /users/{user_id}` with user_id=_bulk and it still landed on the correct handler. 

`/users/_bulk` accepts `List[User]` model, while `users/{user_id]` accepts just `User` model that's the only difference between request to `/users/{user_id}` with user_id=_bulk and request to `/users/_bulk`. I wonder how does FastAPI distinguish between them. Does it try to cast the body? @tiangolo, can you please comment?

**Comment by euri10** [comment]
I just threw a testcase that shows a bug in APIRouter
https://github.com/euri10/fastapi/tree/underscore_route

**Comment by tiangolo** [comment]
So, paths should be evaluated in order.

This section warns about it: https://fastapi.tiangolo.com/tutorial/path-params/#order-matters (that section wasn't always there).

It seems strange if it seemed to work at some point... :thinking:

I just released a version that uses the latest (just released) version of Starlette. I was thinking that it could have some new logic around that. But checking the code, I don't see any change in it...

About the feature request of sorting the routes by specificity, I think it should be in Starlette directly, as FastAPI doesn't override that. FastAPI tries to re-use as much as possible from Starlette, without overriding it, to make the most of all the features already provided by Starlette.

**Comment by haizaar** [comment]
Found the reason why it seemed to work regardless of the order - by accident I defined bulk endpoint as `/users/_bulk/` with trailing `/` while `/users/{user_id}` without.

Such a waste of everyone time :|

Thank you @euri10 and @tiangolo for looking into this.

**Comment by tiangolo** [comment]
Hehe cool. You can declare everything with slashes and it will redirect versions without slashes to versions with slashes.

There are still some cases where it's useful to have non-slashes only routes.

**Comment by haizaar** [comment]
Not exactly - upon 302, client will switch from e.g. POST to get. So `curl -L -X POST /users/_bulk` will actually fail. May be once https://github.com/encode/starlette/issues/464 lands, it will work transparently as long as redirects are enabled in clients.

Though I think I prefer not to rely on redirects, since a). Not all clients, particularly backend clients (svc to svc) enable them by default; b). backend clients will not cache the response (because usually they lack cache all together) so it will result in extra roundtrip to server. I prefer to fail early and explicitly.
What's your experience on the subject?

```
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8000 (#0)
> POST /api/v1/users/_bulk HTTP/1.1
> Host: localhost:5095
> User-Agent: curl/7.47.0
> accept: application/json
> Content-Type: application/json
> Content-Length: 2
> 
* upload completely sent off: 2 out of 2 bytes
< HTTP/1.1 302 Found
< date: Thu, 23 May 2019 11:00:50 GMT
< server: uvicorn
< location: http://localhost:8000/api/v1/users/_bulk/
< transfer-encoding: chunked
< 
* Ignoring the response-body
* Connection #0 to host localhost left intact
* Issue another request to this URL: 'http://localhost:8000/api/v1/pets/_bulk/'
* Switch from POST to GET               # <----------------------------------------!!!!!
...
````

**Comment by tiangolo** [comment]
Hey @haizaar, thanks for your patience.

This should be solved in the latest versions after https://github.com/encode/starlette/pull/563.

**Comment by haizaar** [comment]
/CC @motinani, @worroc

Hi @tiangolo,
I moved to new place where I'm not involved with FastAPI anymore. I'd like to once again express my deep gratitude to you for creating this amazing project!
